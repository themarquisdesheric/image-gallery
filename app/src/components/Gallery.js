import React, { Component } from 'react';
import PropTypes from 'prop-types';
import DeleteButton from './DeleteButton';
import OutOfBunnies from './OutOfBunnies';

class Gallery extends Component {
  constructor(props) {
    super(props);
    
    this.state = { index: 0 };

    this.handleGoToPreviousImage = this.handleGoToPreviousImage.bind(this);
    this.handleGoToNextImage = this.handleGoToNextImage.bind(this);
  }

  static propTypes = { 
    data: PropTypes.array.isRequired,
    onDelete: PropTypes.func.isRequired
  }

  handleGoToPreviousImage() {
    let decrementIndex = this.state.index;
    decrementIndex--;
    const index = decrementIndex === -1 ? this.props.data.length - 1 : decrementIndex;
    this.setState({ index });
  }

  handleGoToNextImage() {
    let incrementIndex = this.state.index;
    incrementIndex++;
    const index = incrementIndex === this.props.data.length ? 0 : incrementIndex;
    this.setState({ index });
  }

  render() {
    let { data, onDelete } = this.props;

    if (!data.length) return <OutOfBunnies />;

    let currentImage = data[this.state.index];

    if (!currentImage) return null;
    
    let { title, description, url, _id } = currentImage;

    return (
      <div className="Gallery">
        <div className="button-container">
          <button onClick={this.handleGoToPreviousImage}>&larr;</button>
          <button onClick={this.handleGoToNextImage}>&rarr;</button>
        </div>
        <div className="Gallery-view">
          <h2>{title}</h2>
          <p>{description}</p>
          <img 
            src={url} 
            alt={title} 
            className="Gallery-image" 
          />
          <br/>
          <DeleteButton 
            onClick={() => {
              onDelete(_id);
            
              if (this.state.index === data.length - 1) this.handleGoToPreviousImage();
            }} 
          />
        </div>
      </div>
    );
  }
}

export default Gallery;