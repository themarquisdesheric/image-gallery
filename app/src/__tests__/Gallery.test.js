import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Gallery from '../components/Gallery';
import images from './_data';

function onDelete() {}

describe('Gallery', () => {
  it('renders a Gallery view for Black & Tan', () => {
    const wrapper = shallow(
      <Gallery data={images} onDelete={onDelete}/>
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders a Gallery view for Falling Bunny', () => {
    const wrapper = shallow(
      <Gallery data={images} onDelete={onDelete}/>
    );

    wrapper.find('button').at(1).simulate('click');  
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders a Gallery view for Marshmellow Puff', () => {
    const wrapper = shallow(
      <Gallery data={images} onDelete={onDelete}/>
    );

    wrapper.find('button').at(1).simulate('click');  
    wrapper.find('button').at(1).simulate('click');  
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders a Gallery view for Floppy Ears', () => {
    const wrapper = shallow(
      <Gallery data={images} onDelete={onDelete}/>
    );

    wrapper.find('button').at(1).simulate('click');  
    wrapper.find('button').at(1).simulate('click');  
    wrapper.find('button').at(1).simulate('click');  
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});