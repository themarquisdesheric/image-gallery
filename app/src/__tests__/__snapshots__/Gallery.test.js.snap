// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery renders a Gallery view for Black & Tan 1`] = `
<div
  className="Gallery"
>
  <div
    className="button-container"
  >
    <button
      onClick={[Function]}
    >
      ←
    </button>
    <button
      onClick={[Function]}
    >
      →
    </button>
  </div>
  <div
    className="Gallery-view"
  >
    <h2>
      Black & Tan
    </h2>
    <p>
      Seriously heart-melting black and tan bunny... OMG &lt;3
    </p>
    <img
      alt="Black & Tan"
      className="Gallery-image"
      src="http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg"
    />
    <br />
    <DeleteButton
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Gallery renders a Gallery view for Falling Bunny 1`] = `
<div
  className="Gallery"
>
  <div
    className="button-container"
  >
    <button
      onClick={[Function]}
    >
      ←
    </button>
    <button
      onClick={[Function]}
    >
      →
    </button>
  </div>
  <div
    className="Gallery-view"
  >
    <h2>
      Falling Bunny
    </h2>
    <p>
      I don't know if this bunny fell and the picture was taken on impact, but I certainly prefer to think of it that way
    </p>
    <img
      alt="Falling Bunny"
      className="Gallery-image"
      src="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-25__605.jpg"
    />
    <br />
    <DeleteButton
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Gallery renders a Gallery view for Floppy Ears 1`] = `
<div
  className="Gallery"
>
  <div
    className="button-container"
  >
    <button
      onClick={[Function]}
    >
      ←
    </button>
    <button
      onClick={[Function]}
    >
      →
    </button>
  </div>
  <div
    className="Gallery-view"
  >
    <h2>
      Black & Tan
    </h2>
    <p>
      Seriously heart-melting black and tan bunny... OMG &lt;3
    </p>
    <img
      alt="Black & Tan"
      className="Gallery-image"
      src="http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg"
    />
    <br />
    <DeleteButton
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Gallery renders a Gallery view for Marshmellow Puff 1`] = `
<div
  className="Gallery"
>
  <div
    className="button-container"
  >
    <button
      onClick={[Function]}
    >
      ←
    </button>
    <button
      onClick={[Function]}
    >
      →
    </button>
  </div>
  <div
    className="Gallery-view"
  >
    <h2>
      Marshmellow Puff
    </h2>
    <p>
      Is it a rodent or a rabbit? I had to look it up to be sure. Then again, I'm still not sure
    </p>
    <img
      alt="Marshmellow Puff"
      className="Gallery-image"
      src="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-110__605.jpg"
    />
    <br />
    <DeleteButton
      onClick={[Function]}
    />
  </div>
</div>
`;
